### ISSUE TITLE
Properties to Yaml config changes list

### ISSUE THREAD

**Issue by lhazlewood** [stormpath/stormpath-sdk-java#588]
We need a compiled list of:

1. The properties that we currently have that are being renamed to a spec-compliant name.
2. The properties that the spec has that we don't currently support.
3. The properties that we currently support that are not in the spec.

**Comment by dogeared** [comment]
Wiki page to capture this information is [here](https://github.com/stormpath/stormpath-sdk-java/wiki/1.0-Configuration-Changes---Additions-Guide)

**Comment by mraible** [comment]
It looks like the wiki link above is wrong. Should be https://github.com/stormpath/stormpath-sdk-java/wiki/1.0-Configuration-Changes-&-Additions-Guide

**Comment by mraible** [comment]
@lhazlewood Where are the spec config values listed? I have a hard time believing that it's this short list:

https://github.com/stormpath/stormpath-sdk-spec/blob/master/specifications/config.md#default-configuration

If it is, do you want me to list all the properties from `default.stormpath.properties` in the "Not In Specification" section of the wiki page?

**Comment by lhazlewood** [comment]
@mraible yes please - we need to see the list of what is spec'd vs what is not to see where any gaps may lie.

We will definitely have properties not in our spec because our spec is a 'lowest common denominator' relevant for config across any language/framework.  There will likely be framework-specific config options as well, but it is really helpful to us and our customers to know the difference.

Also, FWIW, every config option we support (spec or non-spec) needs to be listed in `additional-spring-configuration-metadata.json` files in the Spring Boot support so IDEs can support config discovery and auto-completion, so we'll need to cross-check our compiled list against those files and ensure the files are edited to be up-to-date.

**Comment by lhazlewood** [comment]
@mraible I forgot to list the other specification document.

The one you linked to contains the properties for the core SDK only.  Framework integrations that support webapps also need to support these as well:

https://github.com/stormpath/stormpath-framework-spec/blob/master/example-config.yaml

Core + this file represent the entirety of our config specification.  But again, we will have more than those because of things like spring-specific configs.

**Comment by mraible** [comment]
@lhazlewood One thing I just noticed is that servlet example [stormpath.properties](https://github.com/stormpath/stormpath-sdk-java/blob/master/examples/servlet/src/main/webapp/WEB-INF/stormpath.properties) uses `stormpath.` for its prefix, where as the Core SDK uses `client.`. If you look in `stormpath.properties` from the servlet example, you'll see the following:

```
# stormpath.cache.com.stormpath.sdk.account.Account.tti = 3600000
# stormpath.cache.com.stormpath.sdk.group.Group.ttl = 4800000
#
# etc.


# =========== HTTP Proxy Config ============
# If your app requires HTTP communication to Stormpath to go through an HTTP Proxy:
#
# stormpath.proxy.host = foo
# stormpath.proxy.port = 8888
# stormpath.proxy.username = jsmith
```

Whereas the default `com/stormpath/sdk/config/stormpath.properties` has the following:

```
client.apiKey.file:
client.apiKey.id:
client.apiKey.secret:
client.cacheManager.defaultTtl: 300
client.cacheManager.defaultTti: 300
client.cacheManager.caches:
client.baseUrl: https://api.stormpath.com/v1
client.connectionTimeout: 30
client.authenticationScheme: SAUTHC1
client.proxy.port:
client.proxy.host:
client.proxy.username:
client.proxy.password:
```

For configuring caches, it looks as follows:

```
client.cacheManager.caches.com.stormpath.sdk.impl.client.Account.tti=1000
client.cacheManager.caches.com.stormpath.sdk.impl.client.Account.ttl=150
```

Do we need to fix this discrepancy or just document it?

**Comment by mraible** [comment]
@lhazlewood I might've just found my mistake. In #607, it looks like I need to add `stormpath` before client. Is that correct? If so, it still doesn't solve the problem of `stormpath.proxy.host` versus `stormpath.client.proxy.host`.

**Comment by lhazlewood** [comment]
They didn't document this well, but *all* of our properties - at any level - need to be prefixed with `stormpath.`.  HTH!

**Comment by lhazlewood** [comment]
Per the core sdk spec, it would be `stormpath.client.proxy.host`

**Comment by mraible** [comment]
@lhazlewood I see three different `additional-spring-configuration-metadata.json` files:

1. [extensions/spring/boot/stormpath-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json](https://github.com/stormpath/stormpath-sdk-java/blob/master/extensions/spring/boot/stormpath-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json)
2. [extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json](https://github.com/stormpath/stormpath-sdk-java/blob/master/extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json)
3. [extensions/spring/boot/stormpath-spring-security-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json](https://github.com/stormpath/stormpath-sdk-java/blob/master/extensions/spring/boot/stormpath-spring-security-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json)

They each have groups and properties that are very different. Should all three contain the same information (e.g. properties for all [spec](https://github.com/stormpath/stormpath-framework-spec/blob/master/example-config.yaml) and [default web config](https://github.com/stormpath/stormpath-sdk-java/blob/master/extensions/servlet/src/main/resources/META-INF/com/stormpath/sdk/servlet/default.stormpath.properties) options)?

**Comment by lhazlewood** [comment]
* stormpath-spring-boot-starter: config properties supported by the core SDK (`stormpath.client.`*)

* stormpath-webmvc-spring-boot-starter: config properties for the web tier (all `stormpath.web.`* and `stormpath.application.`* properties) and any others that are specific to that starter (if any)

* stormpath-spring-security-spring-boot-starter: config properties relevant to stormpath-spring-security integration (`stormpath.spring.security.`*)

Because stormpath-spring-boot-starter just makes the core SDK 'spring bootable', only the relevant core SDK properties are necessary.

Simiilarly, stormpath-webmvc-spring-boot-starter makes our web integration 'spring bootable' so you need the SDK properties that are relevant to the web tier.  This *may* include additional properties not in Stormpath's web spec because we might need additional values specific to deploying to Spring Boot.  

In other words, Stormpath's specs are common config properties relevant to *all* languages/frameworks we support.  For any given language framework, there may be additional properties supported if it makes sense for that framework, but the spec properties always need to be supported.

**Comment by mraible** [comment]
Here's some key names that have changed in the spec and need to change for 1.0:

| Current                   | New as of 1.0                |
|---------------------------|------------------------------|
|stormpath.web.idSite.register.uri|stormpath.web.idSite.registerUri|
|stormpath.web.idSite.forgot.uri|stormpath.web.idSite.forgotUri|

The key `stormpath.web.idSite.result.uri` is in default.stormpath.properties, but not in the spec. Should it be renamed to `stormpath.web.idSite.resultUri`?

**Comment by mraible** [comment]
On a related note, the `stormpath.web.saml.result.uri` key is Java-specific and not in the spec. Should it be renamed to be similar to the above keys (e.g. `stormpath.web.saml.resultUri`)?

**Comment by lhazlewood** [comment]
I could be mistaken, but I think the saml stuff is represented in the spec as the [stormpath callback](https://github.com/stormpath/stormpath-framework-spec/blob/master/example-config.yaml#L190-L192).

Any assertion that comes from Stormapth's API - regardless if it is generated from an OAuth or SAML response - should always go to that one callback in the receiving web app.  We made this design decision so that customers using our integration don't have to worry about exposing/securing/configuring multiple URIs - there is only one that is needed.

However, the [social callback uris](https://github.com/stormpath/stormpath-framework-spec/blob/master/example-config.yaml#L201-L213) are endpoints that are invoked directly by the social provider.  They do require exposure in the web app, but these are considered temporary - we still have to support them for the 1.0 spec because that is the way social integration works in Stormpath today.  However, in the future, all social interaction will use the Stormpath API as the middle redirector such that *all* callbacks will eventually be sent to the one stormpathCallback discussed above.  Once that is done and implemented in the REST API, the current social callback configuration options will be removed.  But we need to keep them today because that functionality isn't in the REST API yet.

cc @dogeared @maranai

**Comment by mrioan** [comment]
@mraible what @lhazlewood described is accurate. SAML/IDSite callback uri is defined by `stormpath.web.callback.uri`

**Comment by mraible** [comment]
If I compare `extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json` with `web.stormpath.properties`, I find lots of differences.

First of all, the JSON file has 100 keys and the Properties file has 200.

1. Number of keys that are in `web.stormpath.properties`, that aren't in `additional-spring-configuration-metadata.json`: 143
2. Keys in metadata that aren't in `web.stormpath.properties`: 43

Full listing below.

Keys in web.stormpath.properties that aren't in metadata:
```
stormpath.application.href
stormpath.application.name
stormpath.web.basePath
stormpath.web.oauth2.client_credentials.enabled
stormpath.web.oauth2.client_credentials.accessToken.ttl
stormpath.web.oauth2.password.enabled
stormpath.web.oauth2.password.validationStrategy
stormpath.web.accessTokenCookie.name
stormpath.web.accessTokenCookie.httpOnly
stormpath.web.accessTokenCookie.secure
stormpath.web.accessTokenCookie.path
stormpath.web.accessTokenCookie.domain
stormpath.web.refreshTokenCookie.name
stormpath.web.refreshTokenCookie.httpOnly
stormpath.web.refreshTokenCookie.secure
stormpath.web.refreshTokenCookie.path
stormpath.web.refreshTokenCookie.domain
stormpath.web.produces[0]
stormpath.web.produces[1]
stormpath.web.register.autoLogin
stormpath.web.register.form.fields.givenName.enabled
stormpath.web.register.form.fields.givenName.visible
stormpath.web.register.form.fields.givenName.label
stormpath.web.register.form.fields.givenName.placeholder
stormpath.web.register.form.fields.givenName.required
stormpath.web.register.form.fields.givenName.type
stormpath.web.register.form.fields.middleName.enabled
stormpath.web.register.form.fields.middleName.visible
stormpath.web.register.form.fields.middleName.label
stormpath.web.register.form.fields.middleName.placeholder
stormpath.web.register.form.fields.middleName.required
stormpath.web.register.form.fields.middleName.type
stormpath.web.register.form.fields.surname.enabled
stormpath.web.register.form.fields.surname.visible
stormpath.web.register.form.fields.surname.label
stormpath.web.register.form.fields.surname.placeholder
stormpath.web.register.form.fields.surname.required
stormpath.web.register.form.fields.surname.type
stormpath.web.register.form.fields.username.enabled
stormpath.web.register.form.fields.username.visible
stormpath.web.register.form.fields.username.label
stormpath.web.register.form.fields.username.placeholder
stormpath.web.register.form.fields.username.required
stormpath.web.register.form.fields.username.type
stormpath.web.register.form.fields.email.enabled
stormpath.web.register.form.fields.email.visible
stormpath.web.register.form.fields.email.label
stormpath.web.register.form.fields.email.placeholder
stormpath.web.register.form.fields.email.required
stormpath.web.register.form.fields.email.type
stormpath.web.register.form.fields.password.enabled
stormpath.web.register.form.fields.password.visible
stormpath.web.register.form.fields.password.label
stormpath.web.register.form.fields.password.placeholder
stormpath.web.register.form.fields.password.required
stormpath.web.register.form.fields.password.type
stormpath.web.register.form.fields.confirmPassword.enabled
stormpath.web.register.form.fields.confirmPassword.visible
stormpath.web.register.form.fields.confirmPassword.label
stormpath.web.register.form.fields.confirmPassword.placeholder
stormpath.web.register.form.fields.confirmPassword.required
stormpath.web.register.form.fields.confirmPassword.type
stormpath.web.register.form.fieldOrder[0]
stormpath.web.register.form.fieldOrder[1]
stormpath.web.register.form.fieldOrder[2]
stormpath.web.register.form.fieldOrder[3]
stormpath.web.register.form.fieldOrder[4]
stormpath.web.register.form.fieldOrder[5]
stormpath.web.register.form.fieldOrder[6]
stormpath.web.login.form.fields.login.enabled
stormpath.web.login.form.fields.login.visible
stormpath.web.login.form.fields.login.label
stormpath.web.login.form.fields.login.placeholder
stormpath.web.login.form.fields.login.required
stormpath.web.login.form.fields.login.type
stormpath.web.login.form.fields.password.enabled
stormpath.web.login.form.fields.password.visible
stormpath.web.login.form.fields.password.label
stormpath.web.login.form.fields.password.placeholder
stormpath.web.login.form.fields.password.required
stormpath.web.login.form.fields.password.type
stormpath.web.login.form.fieldOrder[0]
stormpath.web.login.form.fieldOrder[1]
stormpath.web.changePassword.autoLogin
stormpath.web.changePassword.errorUri
stormpath.web.social.facebook.uri
stormpath.web.social.facebook.scope
stormpath.web.social.github.uri
stormpath.web.social.github.scope
stormpath.web.social.google.uri
stormpath.web.social.google.scope
stormpath.web.social.linkedin.uri
stormpath.web.social.linkedin.scope
stormpath.web.me.enabled
stormpath.web.me.uri
stormpath.web.me.expand.apiKeys
stormpath.web.me.expand.applications
stormpath.web.me.expand.customData
stormpath.web.me.expand.directory
stormpath.web.me.expand.groupMemberships
stormpath.web.me.expand.groups
stormpath.web.me.expand.providerData
stormpath.web.me.expand.tenant
stormpath.client.cacheManager
stormpath.web.sendVerificationEmail.uri
stormpath.web.unauthorized.uri
stormpath.web.oauth2.validationStrategy
stormpath.web.request.event.publisher
stormpath.web.request.event.listener
stormpath.web.account.cookie.secure.resolver
stormpath.web.account.session.attribute.names
stormpath.web.account.jwt.resolver
stormpath.web.account.jwt.factory
stormpath.web.account.jwt.signingKey.resolver
stormpath.web.csrf.token.manager
stormpath.client.cacheManager.caches.com.stormpath.sdk.servlet.nonces.ttl
stormpath.web.request.factory
stormpath.web.message.source
stormpath.web.locale.resolver
stormpath.web.account.resolvers
stormpath.web.account.resolvers.header
stormpath.web.account.resolvers.cookie
stormpath.web.account.resolvers.session
stormpath.web.localhost.resolver
stormpath.web.oauth2.authorizer
stormpath.web.oauth2.authorizer.secure.resolver
stormpath.web.oauth2.origin.authorizer.serverUriResolver
stormpath.web.oauth2.authenticationRequestFactory
stormpath.web.oauth2.resultFactory
stormpath.web.refreshToken.authenticationRequestFactory
stormpath.web.refreshToken.resultFactory
stormpath.web.accountStoreResolver
stormpath.web.authc.saver
stormpath.web.authc.savers
stormpath.web.authc.unauthenticatedHandler
stormpath.web.authc.usernamePasswordRequestFactory
stormpath.web.http.authc.schemes
stormpath.web.http.authc.schemes.basic
stormpath.web.http.authc.schemes.bearer
stormpath.web.authz.unauthorizedHandler
stormpath.web.filter.chain.resolver
stormpath.web.uris./assets/css/stormpath.css
stormpath.web.uris./assets/css/custom.stormpath.css
```

Keys in metadata that aren't in web.stormpath.properties:
```
stormpath.web.request.remoteUser
stormpath.web.head
stormpath.web.accessToken
stormpath.web.change
stormpath.web.oauth2.origin.authorizer.originUris
stormpath.web.idSite.forgot
stormpath.web.authc.savers.session.enabled
stormpath.web.register
stormpath.web.request
stormpath.web.idSite
stormpath.web.authc.savers.cookie.enabled
stormpath.web.register.form
stormpath.web.idSite.register
stormpath.web.account.cookie
stormpath.client.proxy
stormpath.web.request.userPrincipal
stormpath.web.stormpathFilter.dispatcherTypes
stormpath.web.account.jwt
stormpath.web.stormpathFilter
stormpath.web.stormpathFilter.enabled
stormpath.web.nonce.cache
stormpath.web.idSite.login
stormpath.client
stormpath.web.head.view
stormpath.web.account.cookie.secure
stormpath.web.forgot
stormpath.web.request.application
stormpath.web.idSite.result
stormpath.web.stormpathFilter.servletNames
stormpath.web.handlerMapping
stormpath.web.verify
stormpath.web.head.cssUris
stormpath.web.request.client
stormpath.web.account
stormpath.web.logout
stormpath.web.stormpathFilter.matchAfter
stormpath.web.head.extraCssUris
stormpath.web
stormpath.web.stormpathFilter.urlPatterns
stormpath.web.handlerMapping.order
stormpath.web.stormpathFilter.order
stormpath.web.login
stormpath.web.head.fragmentSelector
```
